chaos-dashboard panic when to delete workflow by ttcontroller

## Bug Report

**What version of Kubernetes are you using?**
<!-- You can run `kubectl version` -->

1.20

**What version of Chaos Mesh are you using?**
<!-- You can run `kubectl exec -n chaos-mesh {chaos-controller-manager-pod-name} -- /usr/local/bin/chaos-controller-manager -version` -->
2.6.0 
**What did you do? / Minimal Reproducible Example**
<!-- If possible, provide a recipe for reproducing the error. How you installed chaos-mesh. -->

Create an workflow on Chaos Dahsboard 

**What did you expect to see?**

All is OK 

**What did you see instead?** 

```
E0628 01:48:28.900025       1 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 645 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x2ab5e00?, 0x46f1ea0})
	/tmp/go/pkg/mod/k8s.io/apimachinery@v0.26.1/pkg/util/runtime/runtime.go:75 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc002008e80?})
	/tmp/go/pkg/mod/k8s.io/apimachinery@v0.26.1/pkg/util/runtime/runtime.go:49 +0x75
panic({0x2ab5e00, 0x46f1ea0})
	/usr/local/go/src/runtime/panic.go:884 +0x212
github.com/chaos-mesh/chaos-mesh/pkg/dashboard/store/workflow.(*WorkflowStore).DeleteByFinishTime(0xc0001a1400, {0xc0027174a0?, 0x203000?}, 0x2260ff9290000?)
	/home/runner/work/chaos-mesh/chaos-mesh/pkg/dashboard/store/workflow/workflow.go:118 +0xe0
github.com/chaos-mesh/chaos-mesh/pkg/dashboard/ttlcontroller.(*Controller).runWorker(0xc000f6f080)
	/home/runner/work/chaos-mesh/chaos-mesh/pkg/dashboard/ttlcontroller/ttlcontroller.go:77 +0xf0
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x0?)
	/tmp/go/pkg/mod/k8s.io/apimachinery@v0.26.1/pkg/util/wait/wait.go:157 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x0?, {0x357a060, 0xc001098660}, 0x1, 0xc0025143c0)
	/tmp/go/pkg/mod/k8s.io/apimachinery@v0.26.1/pkg/util/wait/wait.go:158 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x274a48a78000, 0x0, 0x0?, 0x0?)
	/tmp/go/pkg/mod/k8s.io/apimachinery@v0.26.1/pkg/util/wait/wait.go:135 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(0x0?, 0x0?, 0x0?)
	/tmp/go/pkg/mod/k8s.io/apimachinery@v0.26.1/pkg/util/wait/wait.go:92 +0x25
created by github.com/chaos-mesh/chaos-mesh/pkg/dashboard/ttlcontroller.Register
	/home/runner/work/chaos-mesh/chaos-mesh/pkg/dashboard/ttlcontroller/ttlcontroller.go:65 +0x155
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x27d4200]
```

**Output of chaosctl**
<!-- If it is related to networkchaos, stresschaos or iochaos, chaosctl would help maintainers better find the bug. -->
<!-- See pkg/chaosctl/README.md for detail -->

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -32,6 +32,7 @@ For more information and how-to, see [RFC: Keep A Changelog](https://github.com/
 
 - Set `replicas: 1` automatically when HA is not enabled [#4079](https://github.com/chaos-mesh/chaos-mesh/pull/4079)
 - Remove redundant where statements [#4084](https://github.com/chaos-mesh/chaos-mesh/pull/4084)
+- Fix dashboard panic when exec delete action by finish_time [#4100](https://github.com/chaos-mesh/chaos-mesh/pull/4100)
 
 ### Security
 

--- a/pkg/dashboard/store/experiment/experiment.go
+++ b/pkg/dashboard/store/experiment/experiment.go
@@ -141,6 +141,10 @@ func (e *experimentStore) DeleteByFinishTime(_ context.Context, ttl time.Duratio
 
 	nowTime := time.Now()
 	for _, exp := range experiments {
+		if exp.FinishTime == nil {
+			log.Error(nil, "finish time is nil when deleting archived experiment records, skip it", "experiment", exp)
+			continue
+		}
 		if exp.FinishTime.Add(ttl).Before(nowTime) {
 			if err := e.db.Table("experiments").Unscoped().Delete(*exp).Error; err != nil {
 				return err

--- a/pkg/dashboard/store/schedule/schedule.go
+++ b/pkg/dashboard/store/schedule/schedule.go
@@ -113,7 +113,7 @@ func (e *ScheduleStore) DeleteByFinishTime(_ context.Context, ttl time.Duration)
 	nowTime := time.Now()
 	for _, sch := range sches {
 		if sch.FinishTime == nil {
-			log.Error(nil, "finish time is nil when deleting schedule records, skip it", "schedule", sch)
+			log.Error(nil, "finish time is nil when deleting archived schedule records, skip it", "schedule", sch)
 			continue
 		}
 		if sch.FinishTime.Add(ttl).Before(nowTime) {

--- a/pkg/dashboard/store/schedule/schedule.go
+++ b/pkg/dashboard/store/schedule/schedule.go
@@ -112,6 +112,10 @@ func (e *ScheduleStore) DeleteByFinishTime(_ context.Context, ttl time.Duration)
 
 	nowTime := time.Now()
 	for _, sch := range sches {
+		if sch.FinishTime == nil {
+			log.Error(nil, "finish time is nil when deleting schedule records, skip it", "schedule", sch)
+			continue
+		}
 		if sch.FinishTime.Add(ttl).Before(nowTime) {
 			if err := e.db.Table("schedules").Unscoped().Delete(*sch).Error; err != nil {
 				return err

--- a/pkg/dashboard/store/workflow/workflow.go
+++ b/pkg/dashboard/store/workflow/workflow.go
@@ -119,7 +119,7 @@ func (it *WorkflowStore) DeleteByFinishTime(ctx context.Context, ttl time.Durati
 	nowTime := time.Now()
 	for _, wfl := range workflows {
 		if wfl.FinishTime == nil {
-			log.Error(nil, "workflow finish time is nil when deleting workflow reocrds, skip it", "workflow", wfl)
+			log.Error(nil, "workflow finish time is nil when deleting archived workflow reocrds, skip it", "workflow", wfl)
 			continue
 		}
 		if wfl.FinishTime.Add(ttl).Before(nowTime) {

--- a/pkg/dashboard/store/workflow/workflow.go
+++ b/pkg/dashboard/store/workflow/workflow.go
@@ -20,9 +20,12 @@ import (
 	"time"
 
 	"github.com/jinzhu/gorm"
+	ctrl "sigs.k8s.io/controller-runtime"
 
 	"github.com/chaos-mesh/chaos-mesh/pkg/dashboard/core"
 )
+
+var log = ctrl.Log.WithName("store/workflow")
 
 type WorkflowStore struct {
 	db *gorm.DB
@@ -115,6 +118,10 @@ func (it *WorkflowStore) DeleteByFinishTime(ctx context.Context, ttl time.Durati
 
 	nowTime := time.Now()
 	for _, wfl := range workflows {
+		if wfl.FinishTime == nil {
+			log.Error(nil, "workflow finish time is nil when deleting workflow reocrds, skip it", "workflow", wfl)
+			continue
+		}
 		if wfl.FinishTime.Add(ttl).Before(nowTime) {
 			if err := it.db.Where("uid = ?", wfl.UID).Unscoped().Delete(*it).Error; err != nil {
 				return err
@@ -128,7 +135,7 @@ func (it *WorkflowStore) DeleteByFinishTime(ctx context.Context, ttl time.Durati
 func (it *WorkflowStore) MarkAsArchived(ctx context.Context, namespace, name string) error {
 	if err := it.db.Model(core.WorkflowEntity{}).
 		Where("namespace = ? AND name = ? AND archived = ?", namespace, name, false).
-		Updates(map[string]interface{}{"archived": true, "end_time": time.Now().Format(time.RFC3339)}).Error; err != nil && !gorm.IsRecordNotFoundError(err) {
+		Updates(map[string]interface{}{"archived": true, "finish_time": time.Now().Format(time.RFC3339)}).Error; err != nil && !gorm.IsRecordNotFoundError(err) {
 		return err
 	}
 	return nil

